controllers/prep_handlers.go

```diff
@@ -173,6 +177,289 @@ func (r *ImageBasedUpgradeReconciler) queryPrecachingStatus(ctx context.Context,
 	return
 }
 
+func getSeedImageMountpoint(seedImage string, output string) (string, error) {
+	var data []interface{}
+	if err := json.Unmarshal([]byte(output), &data); err != nil {
+		return "", fmt.Errorf("failed to unmarshal: %w", err)
+	}
+	for _, img := range data {
+		v := img.(map[string]interface{})
+		repos := v["Repositories"].([]interface{})
+		if len(repos) == 0 {
+			continue
+		}
+		repoName := repos[0].(string)
+		if repoName == seedImage {
+			return v["mountpoint"].(string), nil
+		}
+	}
+	return "", fmt.Errorf("failed to find seedImage mountpoint")
+}
+
+var osReadFile = os.ReadFile
+
+func getBootedStaterootID(path string) (string, error) {
+	data, err := osReadFile(path)
+	if err != nil {
+		return "", fmt.Errorf("failed reading %s: %w", path, err)
+	}
+	var status rpmostreeclient.Status
+	if err := json.Unmarshal(data, &status); err != nil {
+		return "", fmt.Errorf("failed unmarshalling %s: %w", path, err)
+	}
+	for _, deploy := range status.Deployments {
+		if deploy.Booted {
+			return deploy.ID, nil
+		}
+	}
+	return "", fmt.Errorf("failed finding booted stateroot")
+}
+
+func getVersionFromManifest(path string) (string, error) {
+	data, err := osReadFile(path)
+	if err != nil {
+		return "", fmt.Errorf("failed reading %s: %w", path, err)
+	}
+	var v map[string]interface{}
+	if err := json.Unmarshal(data, &v); err != nil {
+		return "", fmt.Errorf("failed to unmarshal %s: %w", path, err)
+	}
+	version, ok := v["version"].(string)
+	if !ok {
+		return "", fmt.Errorf("failed to get version from %s: %w", path, err)
+	}
+	return version, nil
+}
+
+func getKernelArgumentsFromMCOFile(path string) ([]string, error) {
+	data, err := osReadFile(path)
+	if err != nil {
+		return nil, fmt.Errorf("failed reading %s: %w", path, err)
+	}
+	var v map[string]interface{}
+	if err := json.Unmarshal(data, &v); err != nil {
+		return nil, fmt.Errorf("failed to unsmarshal %s: %w", path, err)
+	}
+	spec, ok := v["spec"].(map[string]interface{})
+	if !ok {
+		return nil, fmt.Errorf("failed getting spec in %s", path)
+	}
+	kargs, ok := spec["kernelArguments"].([]interface{})
+	if !ok {
+		return nil, fmt.Errorf("failed getting kernelArguments in %s", path)
+	}
+	args := make([]string, len(kargs)*2)
+	for i, karg := range kargs {
```

Comment:
Please validate kernel args logic before push , after applying full flow verify that MCO operator is not degraded

Reply:
Yes, want to be sure we do as they expect it